Add xcode command to upload sourcemaps automatically (#548)
DataDog/datadog-ci: src/commands/react-native/README.md

You can configure the tool to use Datadog EU by defining the `DATADOG_SITE` environment variable as `datadoghq.eu`. By default, the requests are sent to Datadog US.

> To make these variables available, we recommend to set them in a **encrypted** `datadog-ci.json` file at the root of your project:

Reviewer: ```suggestion
To make these variables available, Datadog recommends setting them in an encrypted `datadog-ci.json` file at the root of your project:
```

| GitHub / GitLab  | https://\<repository-url\>/blob/\<commit-hash\>/\<tracked-file-path\>#L\<line\> |
| Bitbucket | https://\<repository-url\>/src/\<commit-hash\>/\<tracked-file-path\>#lines-\<line\>  |

> ### `xcode`

Reviewer 2: I'm wondering if this section should be at the top of the readme?
It'd be wasteful for users that just use `xcode` to waste time reading the "manual" approach above and only realise later (or miss?) that there's an easier 'xcode' setup.

Or alternatively a table of content at the start could be added so users jump straight to the correct instructions.

WDYT?

Author: I'm ok to do this change, but to reassure you really few people are going to be starting their sourcemaps integration by reading this README.

For a bit of context, debug symbols integration for React Native is hard as you need to do:
- iOS Dsym
- Android mapping files
- JS sourcemaps on iOS builds (this script)
- JS sourcemaps on Android builds

Nearly all other error tracking tools have a wizard that installs everything for you and automates the upload, so I think most RN devs don't expect to upload their sourcemaps manually.

On the docs website, we are going to have 2 sections:

**Automated setup with `datadog-react-native-wizard`**
This section will explain how to run a wizard that will setup all of this. There won't be any mention of `datadog-ci` here and we expect that most of the users will only use this part

**Manual setup**
This section will describe step by step what the wizard does and how you can customize each one to suit your needs best. Here we'll introduce the `react-native xcode` command and mention that it uses `react-native upload` under the hood.

So I believe very few devs will start by looking at what our CI tool has to offer without getting the bigger picture on the docs website :)

Reviewer 2: Cool, maybe we'll also be able to add a link to the 'official doc' in the readme!

The upload only happens when your target has a "Release" build configuration to prevent overwriting existing sourcemaps.

This command can use the same environment variables as the `upload` command: `DATADOG_API_KEY` (required), `DATADOG_SITE` and `DATADOG_SOURCEMAP_INTAKE_URL`.

Then, you need to get the location to your `node` and `yarn` binaries by running in a terminal:

```bash
$ which node #/path/to/node
$ which yarn #/path/to/yarn
```

Then change the "Bundle React Native code and images" build phase in XCode (don't forget to change `/path/to/node` and `/path/to/yarn`):

#### For React Native >= 0.69:

To ensure environment variables are well propagated in the build phase, you first need to create a new `custom-react-native-xcode.sh` file in your `ios` folder.

```bash
#!/bin/sh

REACT_NATIVE_XCODE="node_modules/react-native/scripts/react-native-xcode.sh"
# Replace /opt/homebrew/bin/node (resp. /opt/homebrew/bin/yarn) by the value of $(which node) (resp. $(which yarn))
DATADOG_XCODE="/opt/homebrew/bin/node /opt/homebrew/bin/yarn datadog-ci react-native xcode"

/bin/sh -c "$DATADOG_XCODE $REACT_NATIVE_XCODE" 
```

> This will allow to pass the file's path as one argument to the `with-environment.sh` script in the "Bundle React Native code and images" build phase:

Author: ```suggestion
This allows the file's path to be passed as one argument to the `with-environment.sh` script in the "Bundle React Native code and images" build phase: 
```

| GitHub or GitLab  | https://\<repository-url\>/blob/\<commit-hash\>/\<tracked-file-path\>#L\<line\> |
| Bitbucket | https://\<repository-url\>/src/\<commit-hash\>/\<tracked-file-path\>#lines-\<line\>  |

### `xcode`

> This command can be called from an xcode build phase to execute the react-native bundle command and then upload the sourcemaps.

Reviewer: ```suggestion
This command can be called from an XCode build phase to execute the `react-native bundle` command and upload the source maps.
```

> The upload only happens when your target has a "Release" build configuration to prevent overwriting existing sourcemaps.

Reviewer: ```suggestion

The upload only happens when your target has a "Release" build configuration that prevents overwriting existing source maps.
```

Author: I think people could understand it as:

```
The upload only happens when your target has a "Release" build configuration that contains something that prevents overwriting existing source maps.
```
Whereas what I meant is:
```
The upload only happens when your target has a "Release" build configuration; that prevents overwriting existing source maps when running a build with another configuration such as "Debug".
```
Are you ok with my new formulation?

> This command can use the same environment variables as the `upload` command: `DATADOG_API_KEY` (required), `DATADOG_SITE` and `DATADOG_SOURCEMAP_INTAKE_URL`.

Reviewer: ```suggestion
You can use the same environment variables as the `upload` command: `DATADOG_API_KEY` (required), `DATADOG_SITE`, and `DATADOG_SOURCEMAP_INTAKE_URL`.
```

> Then, you need to get the location to your `node` and `yarn` binaries by running in a terminal:

Reviewer: ```suggestion
To get the location to your `node` and `yarn` binaries, run the following in a terminal:
```

```bash
$ which node #/path/to/node
$ which yarn #/path/to/yarn
```

> Then change the "Bundle React Native code and images" build phase in XCode (don't forget to change `/path/to/node` and `/path/to/yarn`):

Reviewer: ```suggestion
Change the "Bundle React Native code and images" build phase in XCode, including `/path/to/node` and `/path/to/yarn`.
```

Author: I think we can remove this sentence, it was an introduction to the next blocks, but I find it confusing as users might wonder whether they need to do an action at this point (they don't)

#### For React Native >= 0.69:

> To ensure environment variables are well propagated in the build phase, you first need to create a new `custom-react-native-xcode.sh` file in your `ios` folder.

Reviewer: ```suggestion
To ensure environment variables are well propagated in the build phase, you need to create a `custom-react-native-xcode.sh` file in your `ios` folder:
```

```bash
#!/bin/sh

REACT_NATIVE_XCODE="node_modules/react-native/scripts/react-native-xcode.sh"
# Replace /opt/homebrew/bin/node (resp. /opt/homebrew/bin/yarn) by the value of $(which node) (resp. $(which yarn))
DATADOG_XCODE="/opt/homebrew/bin/node /opt/homebrew/bin/yarn datadog-ci react-native xcode"

/bin/sh -c "$DATADOG_XCODE $REACT_NATIVE_XCODE" 
```

> This will allow to pass the file's path as one argument to the `with-environment.sh` script in the "Bundle React Native code and images" build phase:

Reviewer: ```suggestion
This allows the file's path to pass as one argument to the `with-environment.sh` script in the "Bundle React Native code and images" build phase:
```

```bash
set -e
export SOURCEMAP_FILE=./main.jsbundle.map
WITH_ENVIRONMENT="../node_modules/react-native/scripts/xcode/with-environment.sh"
REACT_NATIVE_XCODE="./custom-react-native-xcode.sh" 

/bin/sh -c "$WITH_ENVIRONMENT $REACT_NATIVE_XCODE" 
```

#### For React Native < 0.69:

> Simply change the "Bundle React Native code and images" build phase:

Reviewer: ```suggestion
Change the "Bundle React Native code and images" build phase:
```

```bash
set -e
export SOURCEMAP_FILE=./build/main.jsbundle.map
export NODE_BINARY=node
# Change /opt/homebrew/bin/node (resp. /opt/homebrew/bin/yarn) by the value of $(which node) (resp. $(which yarn))
/opt/homebrew/bin/node /opt/homebrew/bin/yarn datadog-ci react-native xcode node_modules/react-native/scripts/react-native-xcode.sh
```

#### Customize the command

> * The first positional argument is the React Native bundle script. 
> If you use another script that requires arguments, you will need to put this script in a file (e.g. in `scripts/bundle.sh`), then provide the path to this file to the `datadog-ci react-native xcode` command.

Reviewer: ```suggestion
The first positional argument is the React Native bundle script. 

If you use another script that requires arguments, you need to put this script in a file (such as in `scripts/bundle.sh`) and provide this file path in the `datadog-ci react-native xcode` command.
```

✅ Uploaded 1 sourcemap in 0.698 seconds.
✨  Done in 2.50s.
```

### `xcode`

> First, build and link your local datadog-ci package, by running in this directory:

Reviewer: ```suggestion
Build and link your local `datadog-ci` package by running this in the directory:
```

```bash
yarn build
yarn link
```

> Then in your project:

Reviewer: ```suggestion
Run this command in your project:
```

```bash
yarn link @datadog/datadog-ci
```

> If running `yarn datadog-ci` in your project returns `error Command "datadog-ci" not found.`, you can do the following:

Reviewer: ```suggestion
If running `yarn datadog-ci` in your project returns `error Command "datadog-ci" not found.`, run the following:
```

```bash
chmod +x /path/to/datadog-ci/dist/cli.js
cp /path/to/datadog-ci/dist/cli.js /path/to/project/node_modules/.bin/datadog-ci
```

> Then you can follow the usual installation steps.

Reviewer: ```suggestion
Then, follow the usual installation steps.
```

✅ Uploaded 1 sourcemap in 0.698 seconds.
✨  Done in 2.50s.
```

### `xcode`

First, build and link your local datadog-ci package, by running in this directory:

```bash
yarn build
yarn link
```

Then in your project:

```bash
yarn link @datadog/datadog-ci
```

If running `yarn datadog-ci` in your project returns `error Command "datadog-ci" not found.`, you can do the following:

```bash
chmod +x /path/to/datadog-ci/dist/cli.js
cp /path/to/datadog-ci/dist/cli.js /path/to/project/node_modules/.bin/datadog-ci
```

> Then you can follow the usual installation steps

Reviewer 2: ```suggestion
Then you can follow the usual installation steps.
```
